File: backend/services/rating_service.py
from typing import List, Optional, Dict
from sqlalchemy.orm import Session
from sqlalchemy.exc import IntegrityError
from sqlalchemy import func

from backend.models.orm.rating import MenuItemRating, RestaurantFeedback
from backend.models.schemas.rating import MenuItemRatingCreate, RestaurantFeedbackCreate

class RatingService:
    def __init__(self, db: Session):
        self.db = db

    def create_menu_item_rating(self, rating: MenuItemRatingCreate, user_id: int) -> MenuItemRating:
        """Create a new rating for a menu item."""
        try:
            db_rating = MenuItemRating(
                user_id=user_id,
                menu_item_id=rating.menu_item_id,
                rating=rating.rating,
                comment=rating.comment
            )
            self.db.add(db_rating)
            self.db.commit()
            self.db.refresh(db_rating)
            return db_rating
        except IntegrityError:
            self.db.rollback()
            raise ValueError("User has already rated this menu item")

    def get_menu_item_ratings(self, menu_item_id: int) -> List[MenuItemRating]:
        """Get all ratings for a specific menu item."""
        return self.db.query(MenuItemRating).filter(MenuItemRating.menu_item_id == menu_item_id).all()

    def get_user_menu_item_rating(self, user_id: int, menu_item_id: int) -> Optional[MenuItemRating]:
        """Get a user's rating for a specific menu item."""
        return self.db.query(MenuItemRating).filter(
            MenuItemRating.user_id == user_id,
            MenuItemRating.menu_item_id == menu_item_id
        ).first()

    def update_menu_item_rating(self, user_id: int, menu_item_id: int, rating: MenuItemRatingCreate) -> MenuItemRating:
        """Update a user's rating for a menu item."""
        db_rating = self.get_user_menu_item_rating(user_id, menu_item_id)
        if not db_rating:
            raise ValueError("Rating not found")

        db_rating.rating = rating.rating
        db_rating.comment = rating.comment
        self.db.commit()
        self.db.refresh(db_rating)
        return db_rating

    def delete_menu_item_rating(self, user_id: int, menu_item_id: int) -> bool:
        """Delete a user's rating for a menu item."""
        db_rating = self.get_user_menu_item_rating(user_id, menu_item_id)
        if not db_rating:
            return False

        self.db.delete(db_rating)
        self.db.commit()
        return True

    def create_restaurant_feedback(self, feedback: RestaurantFeedbackCreate, user_id: int) -> RestaurantFeedback:
        """Create new restaurant feedback."""
        db_feedback = RestaurantFeedback(
            user_id=user_id,
            feedback_text=feedback.feedback_text,
            category=feedback.category
        )
        self.db.add(db_feedback)
        self.db.commit()
        self.db.refresh(db_feedback)
        return db_feedback

    def get_restaurant_feedback(self, category: Optional[str] = None) -> List[RestaurantFeedback]:
        """Get all restaurant feedback, optionally filtered by category."""
        query = self.db.query(RestaurantFeedback)
        if category:
            if category not in RestaurantFeedback.VALID_CATEGORIES:
                raise ValueError(f"Invalid category. Must be one of: {', '.join(RestaurantFeedback.VALID_CATEGORIES)}")
            query = query.filter(RestaurantFeedback.category == category)
        return query.all()

    def get_user_feedback(self, user_id: int) -> List[RestaurantFeedback]:
        """Get all feedback from a specific user."""
        return self.db.query(RestaurantFeedback).filter(RestaurantFeedback.user_id == user_id).all()

    def get_menu_item_average_rating(self, menu_item_id: int) -> Dict[str, any]:
        """Get the average rating and total number of ratings for a menu item."""
        result = self.db.query(
            func.avg(MenuItemRating.rating).label('average'),
            func.count(MenuItemRating.id).label('total')
        ).filter(
            MenuItemRating.menu_item_id == menu_item_id
        ).first()
        
        return {
            'average_rating': round(float(result.average), 1) if result.average else 0.0,
            'total_ratings': result.total
        }

    def get_menu_items_average_ratings(self, menu_item_ids: List[int]) -> Dict[int, Dict[str, any]]:
        """Get average ratings for multiple menu items at once."""
        results = self.db.query(
            MenuItemRating.menu_item_id,
            func.avg(MenuItemRating.rating).label('average'),
            func.count(MenuItemRating.id).label('total')
        ).filter(
            MenuItemRating.menu_item_id.in_(menu_item_ids)
        ).group_by(
            MenuItemRating.menu_item_id
        ).all()
        
        return {
            r.menu_item_id: {
                'average_rating': round(float(r.average), 1),
                'total_ratings': r.total
            }
            for r in results
        }

File: frontend/src/services/ratingService.ts
import { api } from './api';
import { API_ROUTES } from '../constants/api';

export interface MenuItemRating {
  id: number;
  user_id: number;
  menu_item_id: number;
  rating: number;
  comment?: string;
  created_at: string;
  updated_at: string;
}

export interface RatingAverage {
  average: number;
  total: number;
}

export interface CreateMenuItemRating {
  menu_item_id: number;
  rating: number;
  comment?: string;
}

class RatingService {
  async createOrUpdateRating(menuItemId: number, rating: number, comment?: string): Promise<MenuItemRating> {
    try {
      const response = await api.post(API_ROUTES.ratings.menuItems.create(menuItemId), {
        menu_item_id: menuItemId,
        rating,
        comment: comment || null
      });
      return response.data;
    } catch (error) {
      console.error('Error creating/updating rating:', error);
      throw error;
    }
  }

  async getMenuItemRatings(menuItemId: number): Promise<MenuItemRating[]> {
    try {
      const response = await api.get(`${API_ROUTES.ratings.menuItems.base}/${menuItemId}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching ratings:', error);
      throw error;
    }
  }

  async getUserRating(menuItemId: number): Promise<MenuItemRating | null> {
    try {
      const response = await api.get(API_ROUTES.ratings.menuItems.user(menuItemId));
      return response.data;
    } catch (error: any) {
      if (error.response?.status === 404) {
        return null;
      }
      console.error('Error fetching user rating:', error);
      throw error;
    }
  }

  async deleteRating(menuItemId: number): Promise<void> {
    try {
      await api.delete(`${API_ROUTES.ratings.menuItems.base}/${menuItemId}`);
    } catch (error) {
      console.error('Error deleting rating:', error);
      throw error;
    }
  }

  async getAverageRating(menuItemId: number): Promise<RatingAverage> {
    try {
      const response = await api.get(API_ROUTES.ratings.menuItems.average(menuItemId));
      return response.data;
    } catch (error) {
      console.error('Error fetching average rating:', error);
      return { average: 0, total: 0 };
    }
  }
}

export const ratingService = new RatingService(); 